File: src/saxpy.c
#include "ticktock.h"

void saxpy(int n, float a, float *x, float *restrict y){
  int i;
  #pragma acc kernels
  for (i=0; i<n; i++)
    y[i] = a * x[i] + y[i];
}//end of saxpy


int main(){
  int n=1e7; float a = 5./3.;int i;
  float *x = malloc(sizeof(float)*n);
  float *y = malloc(sizeof(float)*n);
  float temp;
  
  //initialize vectors  
  for (i=0; i<n; i++){
    x[i] = 2.0f;
    y[i] = (i+1.0)*(i-1.0);
  }
  tick_tock tt;
  tick(&tt);
  saxpy(n,a,x,y);
  tock(&tt);
  free(x);
  free(y);
  return 0;
}

File: src/saxpy.f90
program gpu_saxpy

   integer,parameter :: N=1e7
   real,save  :: x(N),y(N),z(N)
   real :: a = 5./3.
   integer it0,it1

   call tick(it0,it1) 
   call saxpy(N,a,x,y,z)
   call tock(it0,it1) 

end program

subroutine saxpy(N,a,x,y,z)
   integer :: N
   real  :: a,x(N),y(N),z(N)

!initialization
   do i = 1,N
      x(i) = real(i)*real(i)
      y(i) = (i+1.)*(i-1.)
   end do

!$acc kernels
   do i = 1,N
      z(i) = x(i)*a+y(i)
   end do
!$acc end kernels

end subroutine
